### PR TITLE
Unmarshal and compare user objects in CompareAndSwapUser

### PR DESCRIPTION
**Pull request by espadolini**
This PR makes it so that `CompareAndSwapUser` does a comparison between unmarshaled user objects rather than comparing the JSON serialization through a backend `CompareAndSwap`. Seeing as this requires fetching the object anyway, the new implementation uses a `ConditionalUpdate` operation, which is generally cheaper.

This fixes the failure condition reported at https://github.com/gravitational/teleport/pull/35325#discussion_r1473524505 , but doesn't implement broader changes (to replace all the `CompareAndSwapUser` calls that almost always already get the current value anyway with calls to `UpdateUser`).

changelog: fix conditional user modifications (used by certain Teleport subsystems such as Device Trust) on users that have previously been locked out due to repeated recovery attempts

### CODE REVIEW

**Comment by espadolini** [comment]
This is barely changelog-worthy, since any modification to the user params item that doesn't go through `CompareAndSwapUser` will also fix the issue.

**Review by rosstimothy — APPROVED** [review]

File: lib/services/local/users.go [hunk 1]
```diff
@@ -441,58 +441,76 @@ func (s *IdentityService) UpsertUser(ctx context.Context, user types.User) (type
...
+	}
 	if err := services.ValidateUser(new); err != nil {
 		return trace.Wrap(err)
 	}
 
-	newRaw, ok := new.WithoutSecrets().(types.User)
+	newWithoutSecrets, ok := new.WithoutSecrets().(types.User)
 	if !ok {
-		return trace.BadParameter("Invalid user type %T", new)
+		return trace.BadParameter("Invalid new user type %T (this is a bug).", new)
```
Comment (codingllama) [inline lib/services/local/users.go:457]: nit:

```suggestion
		return trace.BadParameter("invalid new user type %T (this is a bug)", new)
```

https://go.dev/wiki/CodeReviewComments#error-strings

File: lib/services/local/users.go [hunk 2]
```diff
@@ -441,58 +441,76 @@ func (s *IdentityService) UpsertUser(ctx context.Context, user types.User) (type
...
 	}
-	rev := new.GetRevision()
-	newValue, err := services.MarshalUser(newRaw)
-	if err != nil {
-		return trace.Wrap(err)
+
+	existingWithoutSecrets, ok := existing.WithoutSecrets().(types.User)
+	if !ok {
+		return trace.BadParameter("Invalid existing user type %T (this is a bug).", existing)
```
Comment (codingllama) [inline lib/services/local/users.go:462]: nit:

```suggestion
		return trace.BadParameter("invalid existing user type %T (this is a bug)", existing)
```

Comment (codingllama) [inline lib/services/local/users.go:462]: Also every time I see one of those "this is a bug" messages, I wonder how long it'll take until it appears somewhere public.

File: lib/services/local/users.go [hunk 3]
```diff
@@ -441,58 +441,76 @@ func (s *IdentityService) UpsertUser(ctx context.Context, user types.User) (type
...
-	}
-	existingValue, err := services.MarshalUser(existingRaw)
-	if err != nil {
-		return trace.Wrap(err)
-	}
-	existingItem := backend.Item{
-		Key:   backend.Key(webPrefix, usersPrefix, existing.GetName(), paramsPrefix),
-		Value: existingValue,
-	}
+	const iterationLimit = 5
```
Comment (codingllama) [inline lib/services/local/users.go:472]: What is the rationale behind the retries? Are all ConditionalUpdate callers expected to retry? Why 5 times exactly?

File: lib/services/local/users.go [hunk 4]
```diff
@@ -441,58 +441,76 @@ func (s *IdentityService) UpsertUser(ctx context.Context, user types.User) (type
...
+	for i := 0; i < iterationLimit; i++ {
+		const notWithSecrets = false
```
Comment (codingllama) [inline lib/services/local/users.go:474]: I get it, but this reads a bit weird. Maybe `const withSecrets = false` is a bit better?

File: lib/services/local/users.go [hunk 5]
```diff
@@ -441,58 +441,76 @@ func (s *IdentityService) UpsertUser(ctx context.Context, user types.User) (type
...
+
+		_, err = s.Backend.ConditionalUpdate(ctx, item)
+		if err != nil {
+			if trace.IsCompareFailed(err) {
+				continue
+			}
 			return trace.Wrap(err)
 		}
+
+		if auth := new.GetLocalAuth(); auth != nil {
```
Comment (codingllama) [inline lib/services/local/users.go:505]: Should move this out of the loop, given that the conditional update already passed, for clarity?

**Review by codingllama — COMMENTED** [review]
Thanks for tackling this, Edoardo!

File: lib/services/local/users.go (hunk 4 above)
Comment (espadolini) [reply lib/services/local/users.go:474]: That'd be confusing at the call site tho; I'm not sure why I picked `notWithSecrets` instead of just `withoutSecrets` (which is what I tend to use).

File: lib/services/local/users.go (hunk 3 above)
Comment (espadolini) [reply lib/services/local/users.go:472]: Revisions can change either spuriously (although this doesn't actually in any of our backends right now, outside of insane situations like restoring a backup while the auth server is still running) or because something else modified the item - but from the point of view of a compare-and-swap operation we don't care about that, we only care if the value matches or not, and a concurrent modification that leaves the value identical should still result in a successful compare-and-swap. In practice this isn't actually verified in some of our `CompareAndSwapFoo` implementations, unfortunately: `CompareAndSwapCertAuthority` can fail even though it should succeed if the existing value is concurrently overwritten with a different one that is equivalent after deserialization (in that specific case it's not a big deal because all cases of CASing a cert authority are already on various timed loops to begin with).

tl;dr: conditional updates should always be treated as potentially fallible, and the behavior on a failure should be that you reread the current state again and then make a decision on whether you can and want to keep going, or return an error and let the outer layer deal with that

Oh, and I picked 5 because humans generally have 5 fingers on one hand, although this is such a fringe scenario that I'm going to change it to 3: the initial attempt, one retry because spurious failures could actually happen, and another one because a single retry just feels weird.

File: lib/services/local/users.go (hunk 5 above)
Comment (espadolini) [reply lib/services/local/users.go:505]: I couldn't find a satisfying way to rearrange the loop that wouldn't involve checking the iteration variable against `N-1`; if only Go had Python's `for`-`else` construct... 😢

File: lib/services/local/users.go (hunk 2 above)
Comment (codingllama) [reply lib/services/local/users.go]: No, keep the assertions, they are good. Feel free to ignore this comment, I'm not expecting any changes.

File: lib/services/local/users.go (hunk 3 above)
Comment (codingllama) [reply lib/services/local/users.go]: > and I picked 5 because humans generally have 5 fingers on one hand (...)

That's rock-solid logic, can't argue with that.

Thanks for the explanation. So what I'm gathering is that looped ConditionalUpdates are a practice we'd like to encourage, which is good to know.

Do you think it makes sense to capture the number of retries in a more generic, reusable constant, like `ConditionalUpdateDefaultRetries` (or something to that effect)?

File: lib/services/local/users.go (hunk 5 above)
Comment (codingllama) [reply lib/services/local/users.go:506]: Maybe:

```go
updated := false
for i := 0; i < iterationLimit; i++ {
  // (...)
  if _, err = s.Backend.ConditionalUpdate(ctx, item); err != nil {
    // (...)
  }
  updated = true
  break
}
if !updated {
  return trace.LimitExceeded("failed to update user within %v iterations", iterationLimit)
}

// then do upsertLocalAuthSecrets here
```

Up to you.

**Review by codingllama — APPROVED** [review]

File: lib/services/local/users.go (hunk 3 above)
Comment (espadolini) [reply lib/services/local/users.go]: I don't know if there's a one-size-fits-all value for such a constant, I'm only doing a retry loop very close to the backend here because I want to match the slightly stronger semantics of compare-and-swap.
